### PR TITLE
feat: add runAllTimersAsync from sinonjs

### LINKED ISSUE #1804: Add async method variants from sinonjs/fake-timers to vitest
### Clear and concise description of the problem

fake-timers provides an `await clock.runAllAsync()` call as an alternative to `clock.runAll()` (https://github.com/sinonjs/fake-timers#clockrunall--await-clockrunallasync) which makes it significantly easier to run all timers in a more complex asynchronous context (e.g. timers called by promises that are called by timers). Without it, it can be really difficult to actually have all timers run when those timers are initiated by other promises.

See Stack Overflow discussions such as https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises and https://stackoverflow.com/questions/52177631/jest-timer-and-promise-dont-work-well-settimeout-and-async-function to see scenarios where this is an issue.

For a similar reason, `await clock.tickAsync`, `await clock.nextAsync`, `await clock.runToLastAsync` would ideally also be exposed in some way.

I'm currently having to import sinon-js/fake-timers manually to provide this functionality, and I'd prefer to stay within vitest, especially since it's using fake-timers under the hood.

### Suggested solution

Adding something like:

```ts
runAllTimersAsync(): Promise<void> {
    if (this._checkFakeTimers())
      this._clock.runAllAsync()
}
```

to `packages/vitest/src/integrations/mock/timers.ts` for each associated method (and tests as needed).

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by guillaumeduboc**
close #1804

### CODE REVIEW

**Comment by sheremet-va** [comment]
I see that `throws before allowing infinite recursion` fails on windows 🤔

**Comment by guillaumeduboc** [comment]
> I see that `throws before allowing infinite recursion` fails on windows 🤔

hey @sheremet-va 👋 
I fixed the test

I was thinking I could add `tickAsync`, `nextAsync` and `runToLastAsync` as suggested in the issue
should I do it in this PR as well ?

**Comment by sheremet-va** [comment]
> I was thinking I could add `tickAsync`, `nextAsync` and `runToLastAsync` as suggested in the issue
> should I do it in this PR as well ?

Sure, we can do that. Don't forget about docs though 😄

**Comment by guillaumeduboc** [comment]
Sorry for the delay, my PR is almost done
I stumbled upon an issue with `@sinonjs/fake-timers` as there `runToLastAsync` was responsible for some timeouts in my tests

when https://github.com/sinonjs/fake-timers/pull/451 is merged and published I can bump the `@sinonjs/fake-timers` version and it should be good

**Comment by guillaumeduboc** [comment]
hey @sheremet-va 👋 

Now that the sinonjs/fake-timers issue is fixed, I finished my PR and updated the docs
wdyt ?

File: docs/api/index.md [hunk 1]
```diff
@@ -2964,6 +2992,25 @@ IntersectionObserver === undefined
...
+### vi.runAllTimersAsync
+
+- **Type:** `async() => Promise<Vitest>`
+
+  This method will asynchronously invoke every initiated timer until the timers queue is empty. It means that every timer called during `runAllTimersAsync` will be fired even asynchronous timers. If you have an infinite interval,
+  it will throw after 10 000 tries. For example this will log `2, 1`:
+
+  ```ts
+  setTimeout(async () => {
+    await setTimeout(() => {
```
Comment (sheremet-va) [inline docs/api/index.md:3004]: this has no effect, is it? (unless imported form `timers/promises`, which is not specified in this code block)

File: docs/api/index.md [hunk 2]
```diff
@@ -2533,6 +2533,19 @@ Vitest provides utility functions to help you out through it's **vi** helper. Yo
   vi.advanceTimersByTime(150)
   ```
 
+### vi.advanceTimersByTimeAsync
+
+- **Type:** `async (ms: number) => Promise<Vitest>`
```
Comment (sheremet-va) [inline docs/api/index.md:2538]: Don't need to specify `async`

```suggestion
- **Type:** `(ms: number) => Promise<Vitest>`
```

File: docs/api/index.md [hunk 3]
```diff
@@ -2548,6 +2561,21 @@ Vitest provides utility functions to help you out through it's **vi** helper. Yo
     .advanceTimersToNextTimer() // log 3
   ```
 
+### vi.advanceTimersToNextTimerAsync
+
+- **Type:** `async () => Promise<Vitest>`
```
Comment (sheremet-va) [inline docs/api/index.md:2566]: Don't need to specify `async`

```suggestion
- **Type:** `() => Promise<Vitest>`
```

File: docs/api/index.md [hunk 4]
```diff
@@ -2964,6 +2992,25 @@ IntersectionObserver === undefined
   vi.runAllTimers()
   ```
 
+### vi.runAllTimersAsync
+
+- **Type:** `async() => Promise<Vitest>`
```
Comment (sheremet-va) [inline docs/api/index.md:2997]: Don't need to specify `async`

```suggestion
- **Type:** `() => Promise<Vitest>`
```

File: docs/api/index.md [hunk 5]
```diff
@@ -2977,6 +3024,28 @@ IntersectionObserver === undefined
   vi.runOnlyPendingTimers()
   ```
 
+### vi.runOnlyPendingTimersAsync
+
+- **Type:** `async () => Promise<Vitest>`
```
Comment (sheremet-va) [inline docs/api/index.md:3029]: Don't need to specify `async`

```suggestion
- **Type:** `() => Promise<Vitest>`
```

File: test/core/test/timers.test.ts [hunk 6]
```diff
@@ -639,6 +920,113 @@ describe('FakeTimers', () => {
...
+      global.setTimeout(spies[0], 10)
+      global.setTimeout(spies[1], 50)
+
+      await timers.runOnlyPendingTimersAsync()
+
+      expect(spies[0]).toBeCalled()
+      expect(spies[1]).toBeCalled()
+    })
+
+    it.only('runs all timers in order', async () => {
```
Comment (sheremet-va) [inline test/core/test/timers.test.ts:946]: Please, remove `.only` here 😄

**Review by sheremet-va — CHANGES_REQUESTED** [review]
Looks great! Thank you for your contribution. Have some small questions for docs though.

File: docs/api/index.md (hunk 1 above)
Comment (guillaumeduboc) [reply docs/api/index.md:3004]: what do you mean it has no effect ? the `await` ?

Comment (sheremet-va) [reply docs/api/index.md:3004]: Yes, the await

Comment (guillaumeduboc) [reply docs/api/index.md:3024]: it actually has an impact
I can't find much documentation on this but having a `Promise` in a `setTimeout` has a different behavior from having a synchronous callback

the impact on timers is that this code would not work for `runAllTimers`

Comment (guillaumeduboc) [reply docs/api/index.md:3024]: okay this was not clear to me at all

I've been playing with these a bit and my understanding is that it doesn't change anything in node timers, but it changes in sinonjs/fake-timers

for their synchronous api, an asynchronous callback in a `setTimeout` will block the timer

```typescript
vi.useFakeTimers()

setTimeout(async () => {
  console.log(1)
  await Promise.resolve(true) // this promise will block the timer
  
  console.log(2) // happens after termination
}, 100)

vi.runAllTimers()
```

so you are right, for node it makes no difference, but for our fake timers it does

I will update the example to better reflect on that

**Review by sheremet-va — APPROVED** [review]

**Review by antfu — APPROVED** [review]

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**